### PR TITLE
gitserver: add command LFSSmudge

### PR DESCRIPTION
**Pull request by keegancsmith**
This adds a new command which will run the contents of files through git-lfs-smudge. Smudge is the terminology used by git filters for converting a file from what is stored in git into what you want in your working copy. So what this command will do is fetch the file from LFS.

If the file is not an LFS pointer, git-lfs-smudge ignores it and passes it on. We use the same behaviour in our implementation, but we bypass needing to pass the large file as stdin to gitserver.

We needed to fix LocalGitCommand.StdoutReader since previously it was returning combined output (instead of stdout). We also add a utility which ensures we run against bare clones since this is important to test for git-lfs.

Note: this is still missing other pieces before it can be used properly. In particular we need infrastructure to monitor and prune the LFS cache as well as including git-lfs binary in our images. Additionally I need to test more how authorization works in filling the LFS cache, which means we may need to change this implementation up a bit.

Test Plan: added a test

Part of #42903

### CODE REVIEW

**Comment by keegancsmith** [comment]
meta: what was the purpose of adding `GitLocalCommand`? Previously we would just start up a local gitserver which was super cheap. Now we miss out on a bunch of things which we will only catch at runtime. For example, I just realised I didn't add lfs to the allow list of commands.

**Comment by ryanslade** [comment]
> meta: what was the purpose of adding `GitLocalCommand`? Previously we would just start up a local gitserver which was super cheap. Now we miss out on a bunch of things which we will only catch at runtime. For example, I just realised I didn't add lfs to the allow list of commands.

IIRC it was to allow us to test commands from the `internal/gitserver` package without needing to import `cmd/gitserver/server` since this would cause a cyclical import.

File: internal/gitserver/commands.go [hunk 1]
```diff
@@ -1089,6 +1089,51 @@ func parseDirectoryChildren(dirnames, paths []string) map[string][]string {
...
+	r, err := c.NewFileReader(ctx, repo, commit, path, checker)
+	if err != nil {
+		return nil, err
+	}
+
+	pointer := make([]byte, 201)
+	n, err := io.ReadFull(r, pointer)
+	pointer = pointer[:n]
+	if err == nil {
+		// This file is too big to be an LFS pointer. So we can fallback to
```
Comment (ryanslade) [inline internal/gitserver/commands.go:1114]: What if it's just a small file that isn't a pointer?

**Review by ryanslade — APPROVED** [review]
Is there no way to decide if it's a pointer by looking at the file contents?

**Comment by keegancsmith** [comment]
> > meta: what was the purpose of adding `GitLocalCommand`? Previously we would just start up a local gitserver which was super cheap. Now we miss out on a bunch of things which we will only catch at runtime. For example, I just realised I didn't add lfs to the allow list of commands.
> 
> IIRC it was to allow us to test commands from the `internal/gitserver` package without needing to import `cmd/gitserver/server` since this would cause a cyclical import.

aah ok, that makes sense. I'd consider moving some tests into xtests (ie `package gitserver_test` sort of tests) to avoid the cycle but get more test coverage, but not that important right now.

> Is there no way to decide if it's a pointer by looking at the file contents?

yes you can, but my intention is to only call LFSSmudge for files that we already know are git-lfs files. Given git-lfs-smudge just passes non-pointers through it was simple to just always call it in small file cases. See my test case which ensures both large, small and lfs files all work.

**Pull request merged**